### ISSUE TITLE
Crash in map from ring with inverses

### ISSUE THREAD

**Issue by pzinn** [Macaulay2/M2#2215]
this crashes:
```
i1 : R=QQ[x,Inverses=>true,MonomialOrder=>Lex];

i2 : f=map(QQ,R,{-17})

o2 = map (QQ, R, {-17})

o2 : RingMap QQ <--- R

i3 : f (x^-1)
-- SIGSEGV
-* stack trace, pid: 1019679
M2-binary: malloc.c:2394: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
/home/pzinn/M2/M2/BUILD/fedora/usr-dist/x86_64-Linux-Fedora-33/bin/M2: line 2: 1019679 Aborted                 (core dumped) LD_LIBRARY_PATH=`dirname "$0"`/../lib64/Macaulay2/lib:$LD_LIBRARY_PATH `dirname "$0"`/M2-binary "$@"

Process M2 exited abnormally with code 134
```
incidentally, this also crashes instead of display an error:
```
i1 : R=QQ[x,Inverses=>true,MonomialOrder=>Lex];

i2 : f=map(ZZ,R,{-17})

o2 = map (ZZ, R, {-17})

o2 : RingMap ZZ <--- R

i3 : f (x^-1)
gmp: overflow in mpz type
/home/pzinn/M2/M2/BUILD/fedora/usr-dist/x86_64-Linux-Fedora-33/bin/M2: line 2: 1019812 Aborted                 (core dumped) LD_LIBRARY_PATH=`dirname "$0"`/../lib64/Macaulay2/lib:$LD_LIBRARY_PATH `dirname "$0"`/M2-binary "$@"

Process M2 exited abnormally with code 134
```

**Comment by pzinn** [comment]
wait, am I going crazy? can someone confirm this?
```
i1 : (-17)^(-1)
-- SIGSEGV
-* stack trace, pid: 167473
 0# std::vector<boost::stacktrace::frame, std::allocator<boost::stacktrace::frame> >::size() const at /usr/include/c++/11/bits/stl_vector.h:919
 1# segv_handler at /home/pzinn/M2/M2/BUILD/fedora/../../Macaulay2/bin/main.cpp:244
 2# 0x00007FA4CC445320 in /lib64/libc.so.6
 3# __gmpn_copyi_core2 in /lib64/libgmp.so.10
-- end stack trace *-
```

**Comment by d-torrance** [comment]
> wait, am I going crazy? can someone confirm this?

Yeah, I'm getting the same thing in Debian unstable and Ubuntu 21.04, both with the stable 1.18 release and a recent build of the `development` branch.

It seems like the problem is with odd powers:

```m2
ii1 : (-2)^(-2)

     1
o1 = -
     4

o1 : QQ

i2 : (-2)^(-4)

      1
o2 = --
     16

o2 : QQ

i3 : (-2)^(-6)

      1
o3 = --
     64

o3 : QQ

i4 : (-2)^(-3)
-- SIGSEGV
-* stack trace, pid: 3124595
 0# stack_trace(std::ostream&, bool) at ./M2/Macaulay2/d/main.cpp:124
 1# segv_handler at ./M2/Macaulay2/d/main.cpp:241
 2# 0x00007F0CE5842EF0 in /lib/x86_64-linux-gnu/libc.so.6
 3# __gmpn_copyi in /usr/lib/x86_64-linux-gnu/libgmp.so.10
-- end stack trace *-
```
```

**Comment by mikestillman** [comment]
I'm getting the same thing on MacOS, M2 1.18...  Wow!  I wonder what we (the royal we, likely) screwed up!
@pzinn I've been working on a slightly different fix for inverses (I don't think this one ((-17)^(-1)) would be affected, but I will need to check that).  When the code was first written, it was hard to use exceptions (in older gcc's from 20 years ago, templates and exceptions did not work together, so we did a hack where we returned 0 instead of a value, figuring that we could detect that earlier.

**Comment by DanGrayson** [comment]
I get it, too.
```m2
Macaulay2, version 1.18.0.1
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone

i1 : (-2)^-3
/bin/sh: line 1: 10520 Bus error: 10           ./M2 --no-readline --print-width 103
```

The problem goes back to at least version 1.11.

I can't imagine what it would be.

**Comment by DanGrayson** [comment]
I think I've fixed it in commit c6049b6df9f0d7821e7468eaea52eea7fed5ea6c.  Reopen if not.

**Comment by DanGrayson** [comment]
oops, no, this one should work: commit a77685f1e45014834f94d328505f38bd324ef3b7

**Comment by DanGrayson** [comment]
The original bug about negative odd powers of negative integers goes back to 2003, in commit 1bd38c64e7ab9839868adb9067e687d25e92dfa9.

**Comment by DanGrayson** [comment]
Reopening, as I see I only fixed a secondary bug, not the one first reported, which was this:
```m2
Macaulay2, version 1.18
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone

i1 : R=QQ[x,Inverses=>true,MonomialOrder=>Lex];

i2 : f=map(QQ,R,{-17});

o2 : RingMap QQ <--- R

i3 : f (x^-1);
-- SIGSEGV
```

**Comment by DanGrayson** [comment]
@mikestillman , can you take a look at this?  I think the problem is here:

```c++
R=QQ[x,Inverses=>true,MonomialOrder=>Lex];
f=map(QQ,R,{-17});
f (x^-1);
```

The code gets called with n = -1.

**Comment by DanGrayson** [comment]
... edited the previous comment ...

**Comment by DanGrayson** [comment]
The code looks okay, but why is the assertion there?

**Comment by pzinn** [comment]
more precisely, M2 crashes at aring-qq-gmp.hpp:251 (mpq_set) itself called by aring-glue.hpp:385 (to_ring_elem).

**Comment by mikestillman** [comment]
Fixed in PR #2480.
